src/lib/optimizer/abstract_syntax_tree/abstract_ast_node.hpp

```diff
@@ -218,6 +191,57 @@ class AbstractASTNode : public std::enable_shared_from_this<AbstractASTNode> {
 
   // @{
   /**
+   * Altering the order of Columns in a subtree.
+   *
+   * The order of Columns in a subtree might change, e.g., as the result of Join Re-Ordering. In this call all nodes
+   * above the root of the JoinTree need to adapt the ColumnIDs they refer to. Think, e.g. the ColumnIDs of a SortNode.
+   *
+   * In order to this:
+   *    1. Obtain the ColumnOrigins of the *old* subtree-root by calling get_column_origins() on it.
+   *    2. Hook in the *new* subtree-root.
+   *    3. Call dispatch_column_id_mapping() on the *new* root, passing the ColumnOrigins you obtained in step 1.
+   *
+   * dispatch_column_id_mapping() will
+   *    1. Generate a ColumnIDMapping
+   *    2. Pass this ColumnIDMapping upwards in the tree by calling map_column_ids()
+   *    3. Stop when either the root of the AST is reached OR a Aggregate- or ProjectionNode is reached
+   */
+
+  /**
+   * @returns {get_column_origin(ColumnID{0}), ..., get_column_origin(ColumnID{n-1})}
+   */
+  ColumnOrigins get_column_origins() const;
+
+  /**
+   * For all columns that this node outputs, find the node that originally "created" it, and the ColumnID it is in that
+   * node.
+   * Columns are "created" by Unions, Projections, Aggregates, StoredTables or MockTables, all other nodes just forward
+   * them.
+   *
+   * JoinNode needs to override get_column_origin() - all other nodes will default to forwarding this request to their
+   * left input, *if the column wasn't created by this node*
+   */
+  virtual ColumnOrigin get_column_origin(ColumnID column_id) const;
+
+  /**
+   * See "Altering the order of Columns in a subtree" comment above.
+   *
+   * NOTE: Call this on the node whose Column order changed, NOT on its parents.
```

Comment:
What happens if I call it on the parents? Also: Can we be sure that this is only called if changes have to be made? If yes, do we have an assert that a change is made?